engine/runtime/src/main/java/org/enso/compiler/context/CompilerContext.java

```diff
@@ -0,0 +1,98 @@
+package org.enso.compiler.context;
+
+import com.oracle.truffle.api.TruffleFile;
+import com.oracle.truffle.api.source.Source;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.logging.Level;
+import org.enso.compiler.Cache;
+import org.enso.compiler.Compiler;
+import org.enso.compiler.PackageRepository;
+import org.enso.compiler.core.IR;
+import org.enso.compiler.data.CompilerConfig;
+import org.enso.interpreter.node.ExpressionNode;
+import org.enso.interpreter.runtime.Module;
+import org.enso.interpreter.runtime.scope.ModuleScope;
+import org.enso.interpreter.runtime.scope.TopLevelScope;
+import org.enso.pkg.QualifiedName;
+import org.enso.polyglot.CompilationStage;
+
+/**
+ * Interface that encapsulate all services {@link Compiler} needs from Truffle or other environment.
+ */
+public interface CompilerContext {
+  boolean isIrCachingDisabled();
+
+  boolean isUseGlobalCacheLocations();
+
+  boolean isInteractiveMode();
+
+  PackageRepository getPackageRepository();
+
+  PrintStream getErr();
+
+  PrintStream getOut();
+
+  void log(Level level, String msg, Object... args);
+
+  void logSerializationManager(Level level, String msg, Object... args);
+
+  void notifySerializeModule(QualifiedName moduleName);
+
+  TopLevelScope getTopScope();
+
+  // threads
+  boolean isCreateThreadAllowed();
+
+  Thread createThread(Runnable r);
+
+  Thread createSystemThread(Runnable r);
+
+  // Truffle related
+
+  void truffleRunCodegen(Module module, CompilerConfig config) throws IOException;
+
+  void truffleRunCodegen(Source source, ModuleScope scope, CompilerConfig config, IR.Module ir);
+
+  ExpressionNode truffleRunInline(
+      Source source, InlineContext scope, CompilerConfig config, IR.Expression ir);
+
+  // module related
+  QualifiedName getModuleName(Module module);
+
+  CharSequence getCharacters(Module module) throws IOException;
+
+  void updateModule(Module module, Consumer<Updater> callback);
```

Comment:
This looks good. It will definitelly help us to, at least, make the modifications to modules atomic.